fireflyprotocol/firefly-client-python
# [Feature] Add Tox and Pytest packages for testing and utilities unit tests

This pull request introduces several improvements to the project's testing infrastructure and adds unit tests for utilities. The following changes have been made:

- Tox Configuration: 
   - Created a `tox.ini` file to configure Tox, a testing tool that allows running tests against multiple Python versions and environments.
   - The `tox.ini` file specifies the test environments for Python 3.8 and Python 3.9.
   - Tox will now run the tests on both Python 3.8 and Python 3.9, ensuring compatibility across these versions.

- Pytest Configuration:
   - Created a `pytest.ini` file to configure Pytest, a flexible and powerful testing framework.
   - The `pytest.ini` file contains settings and options for running Pytest.

- Added Tox and Pytest Packages:
   - Included the necessary dependencies for Tox and Pytest in the project.
   - These packages enable efficient and reliable testing procedures.

- Utilities Unit Tests:
   - Added unit tests for utilities to ensure their correctness and functionality.
   - The tests cover various use cases and scenarios, verifying the expected behavior of the utilities.

To run the tests and utilize the newly added testing infrastructure, follow these steps:
- Install Tox and Pytest: 
   - Install Tox and Pytest from `requirements.ini`.

2. Activate the Virtual Environment (Optional):
   - If you prefer to run the tests within a virtual environment, create and activate one using your preferred method.

3. Run the Tests:
   - Run `tox` command in the project's root directory.
   - Tox will execute the tests using both Python 3.8 and Python 3.9 environments defined in the `tox.ini` file.
   - You will see the test results and any reported errors or failures.

Please review and merge this pull request to incorporate these improvements into the project's testing infrastructure.

Sample output of tox
![image](https://github.com/fireflyprotocol/firefly-client-python/assets/16688852/14f3ad93-502c-40d7-acce-f49565ba1545)
